[PHP 8] System ID entropy API

SammyK opened the pull request:
The `zend_system_id` is a (true global) system ID that fingerprints the process state. When extensions add engine hooks during MINIT/startup, entropy is added the system ID for each hook. This allows extensions to identify that changes have been made to the engine since the last PHP process restart. The `zend_system_id` global is finalized and readable as early as `zend_post_startup_cb`.

_The need for this API came about from a bug in OPcache where dangling pointers were introduced into second-level caches after the process state had changed. (See [bug 79825](https://bugs.php.net/bug.php?id=79825) and #5836 for more details.)_

## Adding entropy with engine hooks (MINIT/startup)

The `zend_system_id` global is empty during MINIT/startup to allow extensions to add entropy with each engine hook they use. The system ID is finalized and readable right before the `zend_post_startup_cb` hook is called.

When any of the following overrides are detected after module startup, entropy will be added for each override.

* `zend_compile_file()`
* `zend_execute_ex()`
* `zend_execute_internal()`
* _Are there other overrides we should consider?_

The following engine hook API's were updated to automatically add entropy to the system ID. The signatures had to change slightly to ensure unique entropy could be gathered.

### Reserve handlers

```diff
-ZEND_API int zend_get_resource_handle(zend_extension *extension)
+ZEND_API int zend_get_resource_handle(const char *module_name)
```

Most extensions I've seen just pass garbage into this function ([exhibit A](https://github.com/php/php-src/blob/9d14bdb377171293061f2ae290505a64111ffcb8/ext/opcache/Optimizer/zend_func_info.c#L976-L980), [exhibit B](https://github.com/xdebug/xdebug/blob/b9f3c6fefb98b9e84c6257d2c68a76dc1ee224e3/src/coverage/code_coverage.c#L1002-L1007)) and even if they pass in a proper `zend_extension`, they usually store the handle index in their own global ([exhibit C](https://github.com/DataDog/dd-trace-php/blob/48aae95573544ec678fe1d207bb77773e245a7c7/src/ext/ddtrace.c#L72)).

### Run-time cache slots

```diff
-ZEND_API int zend_get_op_array_extension_handle(void)
+ZEND_API int zend_get_op_array_extension_handle(const char *module_name)
```

_Are there other API's that should be considered?_

To see how these API changes would impact a real-world extension, [here's a patch for Xdebug](https://github.com/xdebug/xdebug/compare/master...SammyK:system-id-api).

## Caveats

Since the system ID is finalized right before the `zend_post_startup_cb` call, any engine hooks added in a `zend_post_startup_cb` hook will not add entropy to the system ID. In these cases, an extension has three options:

1. Move the engine hooks to MINIT/startup if possible.
2. Manually add entropy to the system ID during MINIT/startup for each hook that will occur in `zend_post_startup_cb` (if known at that time).
3. Accept the tradeoff that such a late hook could cause issues with extensions that rely on the system ID.

If option 2 is possible, here's an example of how one might add the entropy manually at MINIT:

```c
static PHP_MINIT_FUNCTION(foo) {
    // Only necessary if these hooks are added later in zend_post_startup_cb
    if (FOO_G(will_hook_resource_handle)) {
        zend_add_system_entropy(PHP_FOO_EXTNAME, "zend_get_resource_handle", NULL, 0);
    }
    if (FOO_G(will_hook_runtime_cache)) {
        zend_add_system_entropy(PHP_FOO_EXTNAME, "zend_get_op_array_extension_handle", NULL, 0);
    }
    if (FOO_G(will_hook_exit_opcode)) {
        zend_uchar opcode = ZEND_EXIT;
        zend_add_system_entropy(PHP_FOO_EXTNAME, "zend_set_user_opcode_handler", &opcode, sizeof(zend_uchar));
    }
    return SUCCESS;
}
```

## Future considerations

`accel_uname_id` could be moved to `zend_system_id.c` but I'm not familiar with Windows builds so I'm not sure how useful this would be.

/cc @dstogov 

## (Removed since original post)

### Custom opcode handlers

```diff
-ZEND_API int zend_set_user_opcode_handler(zend_uchar opcode, user_opcode_handler_t handler);
+ZEND_API int zend_set_user_opcode_handler(zend_uchar opcode, user_opcode_handler_t handler, const char *module_name);
```

dstogov commented on Zend/zend_system_id.c:
```diff
@@ -0,0 +1,79 @@
+
+ZEND_API void zend_add_system_entropy(const char *module_name, const char *hook_name, const void *data, size_t size)
+{
+	if (finalized == 0) {
```
It may make sense to inform about changing something after finalization. I'm not sure...

SammyK replied:
Good point. I'll update it to return a `ZEND_RESULT_CODE`.

nikic replied:
I'd prefer if this was a `ZEND_ASSERT(!finalized)` instead. Seems unlikely anyone is going to check the return value.

SammyK replied:
By design the system ID is finalized right before the `zend_post_startup_cb` hook. Adding `ZEND_ASSERT(!finalized)` will fail for any engine hooks added in a `zend_post_startup_cb` hook (like OPcache does). There are workarounds to add entropy for late hooks (listed in the "Caveats" section of the PR description), but it looks like this design requirement locks us into using the result code approach here.

dstogov commented on main/main.c:
```diff
@@ -2282,6 +2286,9 @@ int php_module_startup(sapi_module_struct *sf, zend_module_entry *additional_mod
 	}
 
+	/* Extensions that add engine hooks after this point do so at their own peril */
+	zend_finalize_system_id();
```
May be call zend_finalize_system_id() from opcache post_startup. I'm also not completely sure. Just an idea...

SammyK replied:
If we moved it there, I'm not sure how we'd be able to finalize the system ID if OPcache is not installed.

SammyK commented:
@dstogov Thanks for the review. I had one more thought over the weekend. We might not have to change the signature of `zend_set_user_opcode_handler`. So long as the same opcodes are overwritten between restarts, the second-level cache files should be stable, right?

If you could help confirm this: The second-level cache file is an issue when a custom handler is provided during compilation but then is later removed. It shouldn't matter which extension provided the custom handler since the serialized op_array from the cache file will just call the handler indexed by the opcode in the `zend_user_opcode_handlers` array.

Example: extension **A** provides a custom opcode handler for `ZEND_EXIT` and the op_array is cached. If the user removes extension **A** and replaces it with extension **B** which also has a custom opcode handler for `ZEND_EXIT`, after a restart the second-level cache file will still work since the index of `zend_user_opcode_handlers` didn't change. Correct?

If that assumption is correct, when finalizing the system ID, we can just iterate over `zend_user_opcode_handlers` and add entropy for any non-NULL opcode handlers. Then we wouldn't need to change the signature of `zend_set_user_opcode_handler`. Or do you think it is still better to add module-specific entropy for `zend_user_opcode_handlers`?

beberlei commented:
@SammyK zend_ast_process hook would be relevant for this

dstogov commented:
@SammyK sorry for delay. I think it's better to remove entropy for user opcode handlers.

SammyK commented:
@dstogov This is ready for another review. I have:

1. Updated `zend_add_system_entropy` to return a `ZEND_RESULT_CODE`
2. Automatically added entropy when there is a `zend_ast_process` override (thanks @beberlei!)
3. Automatically added entropy for all user opcode handlers and reverted the API change to `zend_set_user_opcode_handler`

If you're happy with this solution, perhaps we could get #5836 merged in for PHP 7 as well? :)

dstogov commented:
looks fine, except that this won't catch overrides done in post_startup callback.
I also don't see a big problem with committing simpler solution to PHP-7.4

SammyK commented:
> this won't catch overrides done in post_startup callback.

Correct; that's by design. There needs to be a point in time where the system ID can be used and is guaranteed not to change and `zend_post_startup_cb` is the only startup hook where that's the case. There are some workarounds when hooking that late in startup while still adding entropy to the system ID (described in the "Caveats" section of the PR description).

> I also don't see a big problem with committing simpler solution to PHP-7.4

\o/

SammyK commented:
@sgolemon Should we also consider merging this one in before Tuesday? We'll need to add entropy to the system ID when an observer extension from #5857 is used.

sgolemon commented:
> We'll need to add entropy to the system ID when an observer extension from #5857 is used.

What? Why?

sgolemon commented:
After reading through further, I'd say that the combination of the two does require observers to add entropy, but the observers themselves don't depend on the system_id, so whenever this lands, it can add something to the observer path at that time.  We don't need them to land together for any particular reason.

SammyK commented:
Rebased to resolve conflicts and updated observer call to `zend_get_op_array_extension_handle`.

cmb69 commented:
What's the status here? It seems to me that we should merge this soon, if we want to target 8.0.

SammyK commented:
It should be ready for merge. I can merge in unless anyone has any objections.

SammyK commented:
Just rebased to resolve conflicts after #5836 got merged to master.

cmb69 commented on Zend/zend_system_id.c:
```diff
@@ -0,0 +1,92 @@
+	}
+	PHP_MD5Update(&context, &hooks, sizeof(zend_uchar));
+
+	for (; i < 256; i++) {
```
```suggestion
	for (int16_t i = 0; i < 256; i++) {
```
and remove the definition at the start of the function

cmb69 commented on Zend/zend_system_id.c:
```diff
@@ -0,0 +1,92 @@
+	if (zend_execute_internal) {
+		hooks |= ZEND_HOOK_EXECUTE_INTERNAL;
+	}
+	PHP_MD5Update(&context, &hooks, sizeof(zend_uchar));
```
```suggestion
	PHP_MD5Update(&context, &hooks, sizeof hooks);
```
IMHO cleaner, and more resilient to changes to the type of `hooks`.

cmb69 commented on Zend/zend_system_id.c:
```diff
@@ -0,0 +1,92 @@
+
+	for (; i < 256; i++) {
+		if (zend_get_user_opcode_handler((zend_uchar) i) != NULL) {
+			PHP_MD5Update(&context, &i, sizeof(int16_t));
```
```suggestion
			PHP_MD5Update(&context, &i, sizeof i);
```

cmb69 reviewed (commented):
I haven't thoroughly checked, but LGTM. Maybe address my coding style comments.

nikic reviewed (commented):
Looks reasonable to me.

SammyK commented:
@cmb69 & @nikic Thanks for the code reviews! I've merged in latest master and made tweaks based on the reviews.
